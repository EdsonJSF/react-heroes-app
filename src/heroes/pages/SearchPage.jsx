import { useLocation, useNavigate } from "react-router-dom";
import queryString from "query-string";

import { HeroCard, SearchAlert } from "../components";
import { useForm } from "../../shared";
import { getHeroByName } from "../helpers";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q = "" } = queryString.parse(location.search);
  const heroes = getHeroByName(q);

  const showSearch = {
    show: !q.length,
    info: {
      classStyle: "primary",
      message: "Search a Hero",
      boldMessage: "",
    },
  };
  const showError = {
    show: q.length > 0 && !heroes.length,
    info: {
      classStyle: "danger",
      message: "No Results with",
      boldMessage: q,
    },
  };

  const { search, onInputChange } = useForm({ search: q });

  const handleSearch = (event) => {
    event.preventDefault();

    const findSearch = search.trim().toLowerCase();

    onInputChange({target: {name: "search", value: findSearch}})

    navigate(`?q=${findSearch}`);
  };

  return (
    <>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-12 col-md-5">
          <h4>Searching</h4>
          <hr />

          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Search"
              className="form-control"
              name="search"
              autoComplete="off"
              value={search}
              onChange={onInputChange}
            />
            <button type="submit" className="btn btn-outline-primary mt-1">
              Search
            </button>
          </form>
        </div>

        <div className="col-12 col-md-7">
          <h4>
            Results: <span className="fw-bold">{heroes.length}</span>
          </h4>
          <hr />

          {showSearch.show && <SearchAlert {...showSearch.info} />}
          {showError.show && <SearchAlert {...showError.info} />}

          {heroes.map((hero) => (
            <>
              <HeroCard key={hero.id} hero={hero} />
              <hr />
            </>
          ))}
        </div>
      </div>
    </>
  );
};
