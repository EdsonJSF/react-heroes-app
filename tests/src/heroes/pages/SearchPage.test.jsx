import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { SearchPage } from "../../../../src/heroes/pages/SearchPage";

describe("Test on <SearchPage />", () => {
  test("should show default values", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
