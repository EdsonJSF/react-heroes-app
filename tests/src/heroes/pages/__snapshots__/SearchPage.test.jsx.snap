// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchPage /> should show default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-12 col-md-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="search"
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-12 col-md-7"
    >
      <h4>
        Results: 
        <span
          class="fw-bold"
        >
          0
        </span>
      </h4>
      <hr />
      <div
        class="text-capitalize alert alert-primary"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
